File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.js
function solution(queue1, queue2) {
    const queue = [...queue1, ...queue2];
    let sum1 = queue1.reduce((a, c) => a + c, 0);
    let sum2 = queue2.reduce((a, c) => a + c, 0);
    let goal = (sum1 + sum2) / 2;
    let [start, end, answer] = [0, queue1.length, 0];

    if ((sum1 + sum2) % 2 !== 0) return -1;

    while (start < queue.length && end < queue.length * 2) {
        if (sum1 === goal) return answer;

        if (sum1 < goal) {
            sum1 += queue[end % queue.length];
            end++;
        } else {
            sum1 -= queue[start % queue.length];
            start++;
        }

        answer++;
    }

    return -1;
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        long sum1 = Arrays.stream(queue1).sum();
        long sum2 = Arrays.stream(queue2).sum();
        long goal = (sum1 + sum2) / 2;
        int answer = 0;

        Queue<Integer> arrqueue1 = new LinkedList<>();
        Queue<Integer> arrqueue2 = new LinkedList<>();

        for(int i=0; i<queue1.length; i++) {
            arrqueue1.offer(queue1[i]);
            arrqueue2.offer(queue2[i]);
        }

        while(sum1 != goal) {
            if(sum1 < sum2 && !arrqueue2.isEmpty()) {
                sum1 += arrqueue2.peek();
                sum2 -= arrqueue2.peek();
                arrqueue1.offer(arrqueue2.poll());
            } else if(sum1 > sum2 && !arrqueue1.isEmpty()) {
                sum1 -= arrqueue1.peek();
                sum2 += arrqueue1.peek();
                arrqueue2.offer(arrqueue1.poll());
            }
            answer++;

            if(arrqueue1.size() + arrqueue2.size() + 5 < answer) {
                answer = -1;
                break;
            }
        }

        return answer;
    }
}